scripts/migrate_maas_instance_ids.sh

```diff
@@ -0,0 +1,126 @@
+    #this script with the -r flag for convenience but you should really just
+    #reload the backup that you have created with mongoexport or otherwise if
+    #you run into trouble.
+    mongo_script=$(cat <<'JAVASCRIPT'
+db.instanceData.aggregate([{
+    "$project": {
+        instanceid: {
+            "$cond": {
+                if: {
+                    "$ne": [{
+                        "$substr": ["$instanceid", 0, 20]
+                    }, "/MAAS/api/1.0/nodes/"]
+                }
+                , then: {
+                    "$concat": ["/MAAS/api/1.0/nodes/", "$instanceid", "/"]
+                }
+                , else: "$instanceid"
+            }
+        }
+        , document: "$$ROOT"
+    }
+}, {
+    "$out": "tmp"
+}]);
+ops = db.instanceData.find().map(function (item) {
+    newid = db.tmp.findOne({
+        "document.instanceid": item.instanceid
+    });
+    return {
+        "updateOne": {
+            "filter": {
+                instanceid: newid.document.instanceid
+            }
+            , "update": {
+                "$set": {
+                    instanceid: newid.instanceid
+                }
+            }
+        }
+    }
+});
+db.instanceData.bulkWrite(ops);
+JAVASCRIPT
+)
+else
+    #This script drops the URL wrapping the instance which coerces the data into
+    #MAAS 2.0 format
+    mongo_script=$(cat <<'JAVASCRIPT'
+db.instanceData.aggregate([{
+    "$project": {
+        instanceid: {
+            "$cond": {
+                if: {
+                    "$eq": [{
+                        "$substr": ["$instanceid", 0, 20]
+                    }, "/MAAS/api/1.0/nodes/"]
+                }
+                , then: {
+                    "$substr": ["$instanceid", 20, -1]
+                }
+                , else: "$instanceid"
+            }
+        }
+        , document: "$$ROOT"
+    }
+}, {
+    "$out": "tmp"
+}]);
+ops = db.instanceData.find().map(function (item) {
+    newid = db.tmp.findOne({
+        "document.instanceid": item.instanceid
+    });
+    if (item.instanceid.slice(-1) === "/") {
+        return {
+            "updateOne": {
+                "filter": {
+                    instanceid: newid.document.instanceid
+                }
+                , "update": {
+                    "$set": {
+                        instanceid: newid.instanceid.slice(0, -1)
+                    }
+                }
+            }
+        }
+    } else {
+        return {}
+    }
+});
+db.instanceData.bulkWrite(ops);
+JAVASCRIPT
+)
+fi
+
+#prepare mongoscript for evaluation
+script=$(echo $mongo_script | tr -d ' ' | tr -d '[:blank:]' | tr -d '[:space:]')
+
+agent=$(cd /var/lib/juju/agents; echo machine-*)
+pw=$(sudo grep statepassword /var/lib/juju/agents/${agent}/agent.conf | cut '-d ' -sf2)
+/usr/lib/juju/mongo3.2/bin/mongo --ssl -u ${agent} -p $pw --authenticationDatabase admin --sslAllowInvalidHostnames --sslAllowInvalidCertificates localhost:37017/juju --eval $script
```

Comment:
I believe you can also "| mongo" if you wanted to pass the script content that way.
or maybe "$script" instead of a bare $script might let you pass it as a commandline arg?

Reply:
AFAIK the only ways are to either pass a "real" file to mongo or to pass a single bash word using --eval. Here we choose to use a single bash word to avoid generating a file.

Reply:
I just tested this, I think you just need to use "$script". Specifically I ran:
```
script='db.machines.find({"machineid": "0"}).pretty()'
/usr/lib/juju/mongo3.2/bin/mongo --ssl -u ${agent} -p $pw --authenticationDatabase admin --sslAllowInvalidHostnames --sslAllowInvalidCertificates localhost:37017/juju --eval $script
```
fails because it gets expanded as 2 bash words. but
```
/usr/lib/juju/mongo3.2/bin/mongo --ssl -u ${agent} -p $pw --authenticationDatabase admin --sslAllowInvalidHostnames --sslAllowInvalidCertificates localhost:37017/juju --eval "$script"
```
passes

So even though "$script" embeds " characters its ok, and surrounding it with "" will make it be treated as a single bash word.

Reply:
Just to confirm it, I also did:
```
script=$(cat <<EOF
db.machines.find(
  {
    "machineid": "0"
  }).pretty()
EOF
)
```

And then `"$script"` works while `$script` fails.
You can also see this in effect with just plain `echo $script` vs `echo "$script"`
The former treats all whitespace (including \n) as just a word separator and so it comes out as:
```
$ echo $script
db.machines.find( { "machineid": "0" }).pretty()
$ echo "$script"
db.machines.find(
  {
    "machineid": "0"
  }).pretty()
```

Reply:
I seem to be able to run this bash script with either `"$script"` or simply `$script` alike. I've been running it on xenial against mongo 3.2.